x/dex/contract/runner.go

```diff
@@ -0,0 +1,80 @@
+package contract
+
+import (
+	"sync"
+	"sync/atomic"
+
+	seiutils "github.com/sei-protocol/sei-chain/utils"
+	"github.com/sei-protocol/sei-chain/x/dex/types"
+	"github.com/sei-protocol/sei-chain/x/dex/types/utils"
+)
+
+type ParallelRunner struct {
+	runnable func(contract types.ContractInfo)
+
+	contractAddrToInfo   *sync.Map // of type map[utils.ContractAddress]*types.ContractInfo
+	readyContracts       *sync.Map // of type map[utils.ContractAddress]struct{}
+	readyCnt             int64
+	inProgressCnt        int64
+	someContractFinished chan struct{}
+}
+
+func NewParallelRunner(runnable func(contract types.ContractInfo), contracts []types.ContractInfo) ParallelRunner {
+	contractAddrToInfo := sync.Map{}
+	contractsFrontier := sync.Map{}
+	for _, contract := range contracts {
```

Comment:
The runner will figure out what the next nodes to process within a single `Run()` call. I updated comments for this file and will just paste it here:

```
We define "frontier contract" as a contract which:
1. Has not finished running yet, and
2. either:
   a. has no other contracts depending on it, or
   b. for which all contracts that depend on it have already finished.
Consequently, the set of frontier contracts will mutate throughout the
`Run` method, until all contracts finish their runs.
The key principle here is that at any moment, we can have all frontier
contracts running concurrently, since there must be no ancestral
relationships among them due to the definition above.
The simplest implementation would be:
```
while there is any contract left:
    run all frontier contracts concurrently
    wait for all runs to finish
    update the frontier set
```
We can illustrate why this implementation is not optimal with the following
example:
    Suppose we have four contracts, where A depends on B, and C depends on
    D. The run for A, B, C, D takes 5s, 5s, 8s, 2s, respectively.
    With the implementation above, the first iteration would take 8s since
    it runs A and C, and the second iteration would take 5s since it runs
    B and D. However C doesn't actually need to wait for B to finish, and
    if C runs immediately after A finishes, the whole process would take
    max(5 + 5, 8 + 2) = 10s, which is 3s faster than the implementation
    above.
So we can optimize the implementation to be:
```
while there is any contract left:
    run all frontier contracts concurrently
    wait for any existing run (could be from previous iteration) to finish
    update the frontier set
```
With the example above, the whole process would take 3 iterations:
Iter 1 (A, C run): 5s since it finishes when A finishes
Iter 2 (B run): 3s since it finishes when C finishes
Iter 3 (D run): 2s since it finishes when B, D finish
```

Reply:
I see, so it's improving among the concurrency in the sense that the simple solution is basically sort of like bfs and we process all nodes at level n before moving to level n+1, while the optimized solution is to remove the restriction of processing all contracts at the same level. And this optimization is done by keeping a list which may contain existing contracts when checking frontier contracts each loop. This is awesome, tyty

Reply:
yeah exactly